src/main/java/org/gradle/java/JigsawPlugin.java

```diff
@@ -0,0 +1,231 @@
+
+    private static final String JAVA_PLUGIN = "org.gradle.java";
+
+    private static final String JAVA_LIBRARY_PLUGIN = "org.gradle.java-library";
+
+    private static final String APPLICATION_PLUGIN = "application";
+
+    private static final String EXTENSION_NAME = "architecture";
+
+    private static final String LIBS_PLACEHOLDER = "APP_HOME_LIBS_PLACEHOLDER";
+
+    @Override
+    public void apply(Project project) {
+        LOGGER.debug("Applying JigsawPlugin to " + project.getName());
+        if (!project.getPlugins().hasPlugin(JAVA_PLUGIN) && !project.getPlugins().hasPlugin(JAVA_LIBRARY_PLUGIN)) {
+            project.getPluginManager().apply(JavaPlugin.class);
+        }
+        project.getExtensions().create(EXTENSION_NAME, Architecture.class);
+
+        configureJavaTasks(project);
+    }
+
+    private void configureJavaTasks(final Project project) {
+        project.afterEvaluate(new Action<Project>() {
+            @Override
+            public void execute(final Project project) {
+                configureCompileJavaTask(project);
+                configureCompileTestJavaTask(project);
+                configureJarTask(project);
+                configureTestTask(project);
+                project.getPluginManager().withPlugin(APPLICATION_PLUGIN, new Action<AppliedPlugin>() {
+                    @Override
+                    public void execute(AppliedPlugin appliedPlugin) {
+                        configureRunTask(project);
+                        configureStartScriptsTask(project);
+                    }
+                });
+            }
+        });
+    }
+
+    private void configureCompileJavaTask(final Project project) {
+        final JavaCompile compileJava = (JavaCompile) project.getTasks().findByName(JavaPlugin.COMPILE_JAVA_TASK_NAME);
+        final Architecture module = (Architecture) project.getExtensions().getByName(EXTENSION_NAME);
+        final String moduleDir =
+                compileJava.getDestinationDir().toString() + System.getProperty("path.separator") + module.getModuleName();
+        compileJava.getInputs().property("moduleName", module.getModuleName());
+        compileJava.doFirst(new Action<Task>() {
+            @Override
+            public void execute(Task task) {
+                List<String> args = new ArrayList<>();
+                args.add("--module-path");
+                args.add(compileJava.getClasspath().getAsPath());
+                args.add("-d");
+                args.add(moduleDir);
+                compileJava.getOptions().setCompilerArgs(args);
+                compileJava.setClasspath(project.files());
+            }
+        });
+        compileJava.doLast(new Action<Task>() {
+            @Override
+            public void execute(Task task) {
+                project.copy(new Action<CopySpec>() {
+                    @Override
+                    public void execute(CopySpec copySpec) {
+                        copySpec.from(moduleDir);
```

Comment:
You know what. I'm just totally factually inaccurate. I've made this whole things much more complicated than it needed to be. Thank you for continuing to challenge me. It made me go back to the basics and start checking my assumptions. It turns out that you can use either "a directory with all the classes of a module in it" OR "a directory with another directory that holds a module" as elements on the `--module-path`. So, we can just compile the classes for a project or subproject directly into the classes directory and use that on the module path or the classpath.  This simplifies everything greatly. I'm updating the plugin and the user guide first thing tomorrow.